editor/src/messages/tool/tool_messages/pen_tool.rs

```diff
@@ -236,114 +271,245 @@ impl Fsm for PenToolFsmState {
 						);
 						responses.push_back(add_manipulator_group(
 							&tool_data.path,
+							tool_data.from_start,
 							ManipulatorGroup::new_with_handles(start_position, Some(start_position), Some(start_position)),
 						));
 					}
 
+					// Enter the dragging handle state while the mouse is held down, allowing the user to move the mouse and position the handle
 					PenToolFsmState::DraggingHandle
 				}
 				(PenToolFsmState::PlacingAnchor, PenToolMessage::DragStart) => PenToolFsmState::DraggingHandle,
 				(PenToolFsmState::DraggingHandle, PenToolMessage::DragStop) => {
-					// Add new point onto path
-					if let Some(layer_path) = &tool_data.path {
-						if let Some(manipulator_group) = get_subpath(layer_path, document).and_then(|subpath| subpath.manipulator_groups().last()) {
-							if let Some(out_handle) = &manipulator_group.points[ManipulatorType::OutHandle] {
-								responses.push_back(add_manipulator_group(&tool_data.path, ManipulatorGroup::new_with_anchor(out_handle.position)));
+					let mut process = || {
+						// Get subpath
+						let layer_path = tool_data.path.as_ref()?;
+						let subpath = document.graphene_document.layer(layer_path).ok().and_then(|layer| layer.as_subpath())?;
+
+						// Get the last manipulator group and the one previous to that
+						let mut manipulator_groups = subpath.manipulator_groups().enumerate();
+						let (&last_id, last_manipulator_group) = if tool_data.from_start { manipulator_groups.next()? } else { manipulator_groups.next_back()? };
+						let previous = if tool_data.from_start { manipulator_groups.next() } else { manipulator_groups.next_back() };
+
+						// Get the first manipulator group
+						let mut manipulator_groups = subpath.manipulator_groups().enumerate();
+						let (&first_id, first_manipulator_group) = if tool_data.from_start { manipulator_groups.next_back()? } else { manipulator_groups.next()? };
```

Comment:
I'm noticing we repeat this (and perhaps some of the other code blocks near this section) in a handful of places in this file. Do you see any way we can reduce the code duplication?